### PR TITLE
Enable eprof profiling using config option

### PR DESCRIPTION
**Pull request by djschlegel2**
Make it easy to activate eprof profiling using a basho config param in support of pluggable storage engine diagnostics. 

Bugzid 66882

### CODE REVIEW

File: src/basho_bench_duration.erl [hunk 1]
```diff
@@ -96,20 +96,32 @@ terminate(Reason, #state{duration=DurationMins}) ->
         WSup ->
             WRef = erlang:monitor(process, WSup),
             supervisor:terminate_child(basho_bench_run_sup, basho_bench_worker_sup),
+            supervisor:terminate_child(basho_bench_run_sup, basho_bench_stats),
```
Comment (chewbranca) [inline src/basho_bench_duration.erl]: Is this change intentional?

File: src/basho_bench_duration.erl [hunk 2]
```diff
@@ -96,20 +96,32 @@ terminate(Reason, #state{duration=DurationMins}) ->
...
+            ok;
+        _ ->
+            io:format(standard_error,"Stopping eprof profiling",[]),
+            eprof:stop_profiling(),
+            eprof:log(EprofFile),
+            eprof:analyze(total),
+            io:format(standard_error,"Eprof output in ~p\n",[EprofFile])
+    end,
     supervisor:terminate_child(basho_bench_sup, basho_bench_run_sup),
+    application:set_env(basho_bench_app, is_running, false),
```
Comment (chewbranca) [inline src/basho_bench_duration.erl]: Does moving this line up here make a difference?

File: src/basho_bench_duration.erl (hunk 1 above)
Comment (djschlegel2) [inline src/basho_bench_duration.erl]: I remembered it was helpful in the post-hook stats from my earlier notes but I can remove it for now.

File: src/basho_bench_duration.erl [hunk 3]
```diff
@@ -96,20 +96,32 @@ terminate(Reason, #state{duration=DurationMins}) ->
...
             supervisor:terminate_child(basho_bench_run_sup, basho_bench_worker_sup),
+            supervisor:terminate_child(basho_bench_run_sup, basho_bench_stats),
             receive
                 {'DOWN', WRef, process, _Object, _Info} ->
                     ok
             end
     end,
     run_hook(basho_bench_config:get(post_hook, no_op)),
+    %% If eprof_file is not empty, run eprofiling and write output to filename
+    case EprofFile = basho_bench_config:get(eprof_file, "") of
```
Comment (chewbranca) [inline src/basho_bench_duration.erl]: In general you want to use atoms like `undefined` as default values to match on rather than values like `""`.

The usage of assignment inside the case expression is somewhat odd. It doesn't look like you're using the EprofFile outside of the case clause, so I would just move it in like so:

``` erlang
case basho_bench_config:get(eprof_file, undefined) of
    undefined ->
        ok;
    EprofFile ->
        ...
end
```

File: src/basho_bench_duration.erl [hunk 4]
```diff
@@ -96,20 +96,32 @@ terminate(Reason, #state{duration=DurationMins}) ->
...
+    %% If eprof_file is not empty, run eprofiling and write output to filename
+    case EprofFile = basho_bench_config:get(eprof_file, "") of 
+        "" ->
+            ok;
+        _ ->
+            io:format(standard_error,"Stopping eprof profiling",[]),
+            eprof:stop_profiling(),
+            eprof:log(EprofFile),
+            eprof:analyze(total),
+            io:format(standard_error,"Eprof output in ~p\n",[EprofFile])
```
Comment (chewbranca) [inline src/basho_bench_duration.erl]: Let's use the builtin logging macros here rather than `io:format`.

File: src/basho_bench_duration.erl (hunk 2 above)
Comment (djschlegel2) [inline src/basho_bench_duration.erl]: From printfs I was finding that it never gets past terminate the basho_bench_run_sup in any event. That may be an artifact of my current runs. Originally the eprof-stop/output code was in basho_bench_app:stop but I found it wasn't making it that far even when I included call to stop things. I can poke at that issue a little more.

**Comment by chewbranca** [comment]
Do we want to allow the user to specify the eprof log file? Seems like we could just use a boolean `enable_eprof` and then write to `eprof.log` in the results dir.

File: src/basho_bench_duration.erl [hunk 5]
```diff
@@ -102,6 +102,17 @@ terminate(Reason, #state{duration=DurationMins}) ->
             end
     end,
     run_hook(basho_bench_config:get(post_hook, no_op)),
+    case basho_bench_config:get(enable_eprof, false) of 
+        false ->
+            ok;
+        true ->
+            ?CONSOLE("Stopping eprof profiling",[]),
+            EprofFile = filename:join(basho_bench:get_test_dir(),"eprof.log"),
```
Comment (chewbranca) [inline src/basho_bench_duration.erl]: Should be a space after the comma and before the quote in `"epr...`.

File: src/basho_bench_duration.erl [hunk 6]
```diff
@@ -102,6 +102,17 @@ terminate(Reason, #state{duration=DurationMins}) ->
...
+            eprof:stop_profiling(),
+            eprof:log(EprofFile),
+            eprof:analyze(total),
+            ?CONSOLE("Eprof output in ~p\n",[EprofFile])
```
Comment (chewbranca) [inline src/basho_bench_duration.erl]: Should be spaces after the comma in this line and the other `?CONSOLE` call above.

**Comment by chewbranca** [comment]
+1 with minor whitespace nits sorted.
